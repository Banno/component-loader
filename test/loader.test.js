/* eslint no-undefined: "off", no-useless-escape: "off" */

import loader from '../src';

const normalisePaths = result => result.replace('src\\\\', 'src/');

describe('loader', () => {
  let opts;

  beforeEach(() => {
    opts = {
      callback: jest.fn(),
      resourcePath: 'src/test.html',
      query: {},
    };
  });

  test('can process basic input', () => {
    loader.call(opts, '<div></div>');

    const [call] = opts.callback.mock.calls;

    expect(call[0]).toBe(null);
    expect(normalisePaths(call[1])).toMatchSnapshot();
    expect(call[2]).toBe(undefined);
  });

  test('can process without options', () => {
    opts.query = null;

    loader.call(opts, '<div></div>');

    const [call] = opts.callback.mock.calls;

    expect(call[0]).toBe(null);
    expect(normalisePaths(call[1])).toMatchSnapshot();
    expect(call[2]).toBe(undefined);
  });

  describe('links', () => {
    test('transforms links', () => {
      loader.call(opts, '<link rel="import" href="foo.html">');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('ignores links with invalid href', () => {
      loader.call(opts, '<link rel="import" href="">');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('ignoreLinks option', () => {
      opts.query.ignoreLinks = ['foo.html'];

      loader.call(opts, '<link rel="import" href="foo.html">' +
        '<link rel="import" href="foofoo.html">');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('ignoreLinksFromPartialMatches option', () => {
      opts.query.ignoreLinksFromPartialMatches = ['foo.html'];

      loader.call(opts, '<link rel="import" href="foo.html">' +
        '<link rel="import" href="foofoo.html">');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('ignorePathReWrite option', () => {
      opts.query.ignorePathReWrite = ['foo.html'];

      loader.call(opts, '<link rel="import" href="foo.html">' +
        '<link rel="import" href="foofoo.html">');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });
  });

  describe('domModule', () => {
    test('transforms dom-modules', () => {
      loader.call(opts, '<dom-module id="x-foo">' +
        '<div></div></dom-module>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('ignores invalid HTML', () => {
      loader.call(opts, '</td>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('removes script tags without a source', () => {
      loader.call(opts, '<dom-module id="x-foo">' +
        '<script>var x = 1;</script></dom-module>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).not.toBe(undefined);
    });

    test('removes script tags without a protocol', () => {
      loader.call(opts, '<dom-module id="x-foo">' +
        '<script src="foo.js"></script></dom-module>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('removes link tags', () => {
      loader.call(opts, '<link rel="import" href="test.html">' +
        '<dom-module id="x-foo"></dom-module>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('adds to body if no dom-module', () => {
      loader.call(opts, '<span></span>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });
  });

  describe('scripts', () => {
    test('transforms scripts with a source into imports', () => {
      loader.call(opts, '<script src="foo.js"></script>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('maintains external scripts', () => {
      loader.call(opts, '<script src="http://example.com/test.js"></script>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).toBe(undefined);
    });

    test('maintains inline scripts', () => {
      loader.call(opts, '<script>var x = 5;</script>');

      const [call] = opts.callback.mock.calls;
      expect(call[0]).toBe(null);
      expect(normalisePaths(call[1])).toMatchSnapshot();
      expect(call[2]).not.toBe(undefined);
    });
  });
});
