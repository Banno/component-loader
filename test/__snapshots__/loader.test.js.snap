// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader can process basic input 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.toBody('<div></div>');
"
`;

exports[`loader can process without options 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.toBody('<div></div>');
"
`;

exports[`loader domModule adds to body if no dom-module 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.toBody('<span></span>');
"
`;

exports[`loader domModule ignores invalid HTML 1`] = `""`;

exports[`loader domModule removes link tags 1`] = `
"
import './test.html';

const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.register('<dom-module id=\\"x-foo\\"></dom-module>');
"
`;

exports[`loader domModule removes script tags without a protocol 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.register('<dom-module id=\\"x-foo\\"></dom-module>');

import './foo.js';
"
`;

exports[`loader domModule removes script tags without a source 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.register('<dom-module id=\\"x-foo\\"></dom-module>');

var x = 1;
"
`;

exports[`loader domModule transforms dom-modules 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.register('<dom-module id=\\"x-foo\\"><div></div></dom-module>');
"
`;

exports[`loader links ignoreLinks option 1`] = `
"
import './foofoo.html';
"
`;

exports[`loader links ignoreLinksFromPartialMatches option 1`] = `""`;

exports[`loader links ignorePathReWrite option 1`] = `
"
import 'foo.html';

import 'foofoo.html';
"
`;

exports[`loader links ignores links with invalid href 1`] = `""`;

exports[`loader links transforms links 1`] = `
"
import './foo.html';
"
`;

exports[`loader scripts maintains external scripts 1`] = `
"
const RegisterHtmlTemplate = require('polymer-webpack-loader/register-html-template');
RegisterHtmlTemplate.toBody('<script src=\\"http://example.com/test.js\\"></script>');
"
`;

exports[`loader scripts maintains inline scripts 1`] = `
"
var x = 5;
"
`;

exports[`loader scripts transforms scripts with a source into imports 1`] = `
"
import './foo.js';
"
`;
